src/LanguageServer/Impl/Sources/DefinitionSource.cs

```diff
@@ -46,71 +46,146 @@ internal sealed class DefinitionSource {
+            definingMember = null;
 
-            var m = analysis.ExpressionEvaluator.LookupNameInScopes(name, out var scope);
-            if (m != null && scope.Variables[name] is IVariable v) {
-                member = v;
-                var definition = v.Definition;
-                if (statement is ImportStatement || statement is FromImportStatement) {
-                    // If we are on the variable definition in this module,
-                    // then goto definition should go to the parent, if any.
-                    var indexSpan = v.Definition.Span.ToIndexSpan(analysis.Ast);
-                    var index = location.ToIndex(analysis.Ast);
-                    if (indexSpan.Start <= index && index < indexSpan.End) {
-                        if (v.Parent == null) {
-                            return null;
-                        }
-                        definition = v.Parent.Definition;
-                    }
-                }
+            var name = (expr as NameExpression)?.Name;
+            if (string.IsNullOrEmpty(name)) {
+                return null;
+            }
 
-                if (CanNavigateToModule(definition.DocumentUri)) {
-                    return new Reference { range = definition.Span, uri = definition.DocumentUri };
+            var index = statement.Names.IndexOf(x => x?.MakeString() == name);
+            if (index < 0) {
+                return null;
+            }
+
+            var module = analysis.Document.Interpreter.ModuleResolution.GetImportedModule(name);
+            if (module != null) {
+                definingMember = module;
+                return new Reference { range = default, uri = CanNavigateToModule(module) ? module.Uri : null };
+            }
+
+            // Import A as B
+            if (index >= 0 && index < statement.AsNames.Count) {
+                var value = analysis.ExpressionEvaluator.GetValueFromExpression(statement.AsNames[index]);
+                if (!value.IsUnknown()) {
+                    definingMember = value as ILocatedMember;
+                    return FromMember(definingMember);
                 }
             }
             return null;
         }
 
-        private Reference FromMemberExpression(MemberExpression mex, IDocumentAnalysis analysis, out ILocatedMember member) {
-            member = null;
+
+        private Reference TryFromVariable(string name, IDocumentAnalysis analysis, SourceLocation location, Node statement, out ILocatedMember definingMember) {
+            definingMember = null;
+
+            var m = analysis.ExpressionEvaluator.LookupNameInScopes(name, out var scope);
+            if (m == null || !(scope.Variables[name] is IVariable v)) {
+                return null;
+            }
+
+            definingMember = v;
+            if (statement is ImportStatement || statement is FromImportStatement) {
```

Comment:
Do we care about to what part of the statement `location` belongs to?

Reply:
If it gets here it is a variable of some sort like `C` in `from A import B as C`. We do care below if we are on a variable which has parent (aka linked from another module) so we can go there.